### ISSUE TITLE
Vision Service: Use an inference server or framework instead of custom microservices

### ISSUE THREAD

**Issue by vigsterkr** [photoprism/photoprism-vision#1]
first of all kudos for photoprism! I think it is a great idea to factor out the ML part to a microservice, but i'm really wondering whether it is a good idea to develop an ML server from scratch. Especially looking at the current code, this one hard-codes models, and that is really not something that makes things scalable. there are plenty of mature ML serving development out there, TF has it's own `TensorFlow Serving` but that of course has its own problem, meaning it ties you to use a specific backend, namely TF.

https://github.com/roboflow/inference is something that comes to mind that is backend agnostic, but here's a quite good list of possible options: https://github.com/topics/inference-server

**Comment by lastzero** [comment]
My apologies for not getting back to you sooner! This is just a first version, and we wanted to integrate what we already have before making further changes to the vision service:

- https://github.com/photoprism/photoprism/issues/1090
- https://github.com/photoprism/photoprism/issues/127#issuecomment-2781700031

Now that this is almost done, we can move on to improving the API implementation and models in this repository, e.g. by building on an existing inference server or framework. Any help with this would be very much appreciated! 🥇

**Comment by sgflt** [comment]
Just architecture question.

Is this module designed as proxy for photoprism/anyapp to other possible apis (roboflow, ollama, localai), or is it competetive API, or both?

```
photoprism -> photoprism-vision -+-> (other providers)
                                 |
                                 +-> local tensorflow
```

or

```
photoprism  -+-> (other providers)
             |
             +-> photoprism-vision     
```                               
or
```
photoprism -+-> (other providers) 
            | 
            +-> photoprism-vision -+-> (other providers)
                                   |
                                   +-> local tensorflow

**Comment by lastzero** [comment]
With the latest update, PhotoPrism (see main repository and release notes) can be configured to act as a model proxy, server, and client.

However, the API / server in this repository does not act as a proxy and using Python for this, when we can do it with Go (which we already have) probably makes no sense. We could however refactor the API to use a framework/server software specialized in this, and/or support standardized request and response formats, if there are any (we haven't found any suitable standards when we checked last year, but also didn't spend a lot of time doing research).

**Comment by sgflt** [comment]
I have created draft PR #5. May I ask you for quick review, whether it meets your requirements?

**Comment by lastzero** [comment]
@sgflt Thanks, that's awesome! As I understand it, your PR adds support for generating captions with the Ollama model? 

Note that for this to be testable/usable with the newly added `photoprism vision run` command, (at least) the existing service endpoints will need to be modified to also support data URLs (since the CLI command can't directly share images over https, which requires a running web server accessible from the vision service):

- https://github.com/photoprism/photoprism-vision/issues/4

If data URLs are supported, we will be happy to quickly test and then merge your changes! 👍 

Related Discussion and API Documentation:

- https://github.com/photoprism/photoprism/discussions/4983
- https://docs.photoprism.dev/#/Vision

**Comment by sgflt** [comment]
Yes. I have deduplicated endpoints to single one that routes request by "model" to ImageProcessor that can handle requested model.

OllamaImageProcessor fetches all available models on ollama instance so routes are dynamically determined at start now.

Implementation of #4 was necessary requirement, so I did it too.

TODO:
- [x] Lazy loading of integrations. All models and integrations are loaded at start.
- [x] nsfw api
- [ ] faces api
- [x] support array of images (currently only first image in array is analyzed)


Also I have encountered problem with parsing of models that have slash in their name. It would be probably better to pass model as part of JSON body or query parameter instead of path variable.

**Comment by sgflt** [comment]
I am not sure how each single image is identified in a batch request.

```go
if apiRequest, err = NewApiRequest(model.EndpointRequestFormat(), images, model.EndpointFileScheme()); err != nil

...
for _, label := range apiResponse.Result.Labels {
	result = append(result, label.ToClassify())
}
```

Or is there guarantee that there will be only single image in ApiRequest even if the type is array? 

```go
type Files = []string
type ApiRequest struct {
	Id             string    `form:"id" yaml:"Id,omitempty" json:"id,omitempty"`
	Model          string    `form:"model" yaml:"Model,omitempty" json:"model,omitempty"`
	Url            string    `form:"url" yaml:"Url,omitempty" json:"url,omitempty"`
	Images         Files     `form:"images" yaml:"Images,omitempty" json:"images,omitempty"`
	responseFormat ApiFormat `form:"-"`
}
```

**Comment by lastzero** [comment]
@sgflt Multiple 224x224 thumbnails of an image may be passed to the `labels` service, unless the aspect ratio is 1:1 (i.e. the image is square), see the following overview of thumbnail types:

- https://docs.photoprism.app/developer-guide/media/thumbnails/#standard-sizes

This means that if you have a portrait or landscape image, the thumbnails `left_224`, `right_224` and `tile_224` (center) will be passed and the matching labels will then be merged to avoid duplicates:

https://github.com/photoprism/photoprism/blob/79b6a757039be4ce60c72ed39b813084a807555d/internal/ai/vision/labels.go#L45-L63

From what I know, some frameworks/models can do this on their own (generate multiple square thumbs from a larger image), but in our case, the thumbs are (already) generated and passed individually. This can be changed if needed, but we'd need to know that it works before we do it.

**Comment by sgflt** [comment]
OK, I have adapted code. Now it computes labels from all available images.

I though I can do labelig and embedding within this PR, but my knowledge doesn't allows further progress, so I am probably done.

**Comment by lastzero** [comment]
Thank you very much! I'll try to test and merge it in the next few days, and you can always open a new PR for additional changes.

**Comment by lastzero** [comment]
@sgflt Thanks again for your contribution! We really appreciate it.

Note that I'm leaving this open because it's an "epic" and its intended scope goes beyond the changes we've just merged. What additional architectural changes should ultimately be implemented is still up for debate:

I don't think the service should be too generic (and thus potentially require a lot of configuration), given the intended purpose of allowing users to easily use additional models? So would it really be a good idea to build this service on top of https://github.com/roboflow/inference, as the issue description *seems* to suggest, or would it make more sense to integrate it in a similar way as we've just integrated https://github.com/ollama/ollama? 🤔

**Comment by sgflt** [comment]
Roboflow seems to be framework with its own frontend built on top of [ultralytics](https://github.com/ultralytics/ultralytics).

I have already  tried to integrate roboflow as Ollama with their API client but example failed on authorization to their cloud api even in self hosted container ❔❗ 


**Technical subjective opinion**
Chosen **Adapter Proxy** approach provides sensible defaults and adapts standard apis to custom ones that may fit better to required use case. That may include additional tasks (filtering, resizing, etc.) on data returned from remote API.

Building this service on top of other specific code would be like duplication of that given service itself.

**Comment by sgflt** [comment]
I would sum it up as following:
If it does have an API, use API client. (roboflow)
If it is a library, it may be better to use it as another LocalImageProcessor. (ultralytics)

**Comment by sgflt** [comment]
FYI I am working on direct integration of ultralytics.

**Comment by lastzero** [comment]
@sgflt I now have ollama running locally with the model "llama3" loaded to test your changes before building a new docker image.  Using it in a terminal via "docker exec -it ollama ollama run llama3" works fine, and the pre-existing kosmos-3 model also works, e.g. when I run "photoprism vision run -m caption year:2025" (in the photoprism service container, which can connect to the photoprism-vision service via HTTP).

However, I found that the default `vision/caption` endpoint doesn't check the "model" or "version" passed with the request. So I've been trying to figure out which (sub-)endpoint to call to generate image captions with the ollama service. The `vision.yml` file used for configuration looks like this (only the relevant parts for brevity):

```yaml
- Type: caption
  Resolution: 224
  Name: ollama
  Version: llama3
  Service:
    Uri: http://192.168.1.4:5000/api/v1/vision/caption/ollama
    FileScheme: data
    RequestFormat: vision
    ResponseFormat: vision
```

The endpoint paths `vision/caption/ollama`, `vision/caption/llama3` and `vision/caption/ollama/llama3` all don't seem to work and fail e.g. with the error message "There is no image processor that has ollama available". Of course, it could also be a problem with the environment variables I set to configure the ollama service, and I have already tried "llama3" and "ollama" as the "model" name in the POST request, for example:

```json
{
  "id": "7a0ccf20-981c-4672-8b45-cc280eaa794e",
  "model": "ollama",
  "version": "llama3",
  "images": [
    "data:image/jpeg;..."
  ]
}
```

As I have very little time available for testing, could you please share the exact **endpoint path, method and request values** to use, so I can complete this and add an ollama request example to the docs for future reference?

It might also be helpful to know **which ollama model you used for testing** and if there are any other commands that need to be run in the ollama container to make this work?

What error should I see in the photoprism-vision logs and/or service response if the ollama processor is responding but the ollama service isn't reachable? Does it check/require the name of the model I've loaded in ollama?

Thank you very much! 🙏

**Comment by lastzero** [comment]
Has anyone taken a look at the Docker Model Runner yet, see https://docs.docker.com/model-runner/? It's in beta, and although they call it a plugin, it seems to only be available on Docker Desktop for Mac with Apple Silicon or Windows with NVIDIA GPUs. This would unfortunately make it unsuitable for use with PhotoPrism on a Linux server... 🐳 🚫

**Comment by sgflt** [comment]
I have following configuration:

```yaml
Models:
- Type: nsfw
  Resolution: 224
  Service:
    Uri: http://photoprism-vision:5000/api/v1/vision/nsfw/nsfw_image_detector
    FileScheme: data
    RequestFormat: vision
    ResponseFormat: vision

- Type: labels
  Resolution: 224
  Service:
    Uri: http://photoprism-vision:5000/api/v1/vision/labels/minicpm-v:latest
    FileScheme: data
    RequestFormat: vision
    ResponseFormat: vision

- Type: caption
  Resolution: 224
  Service:
    Uri: http://photoprism-vision:5000/api/v1/vision/caption/minicpm-v:latest
    FileScheme: data
    RequestFormat: vision
    ResponseFormat: vision
```


**minicpm-v:latest** is name of model from first column `docker exec ollama ollama list` including the tag. 

```
ollama list
NAME                                          ID              SIZE      MODIFIED         
minicpm-v:latest                              c92bfad01205    5.5 GB    3 days ago      
llava-phi3:latest                             c7edd7b87593    2.9 GB    3 days ago      
gemma3:4b-it-qat                              d01ad0579247    4.0 GB    3 days ago      
gemma3:12b-it-qat                             5d4fa005e7bb    8.9 GB    4 days ago      
gemma3:27b-it-qat                             29eb0b9aeda3    18 GB     4 days ago 
```

Maybe it would be better to parse model and version only from JSON body, because if model contains `/` it will fail on some parsing error.


I have used minicpm-v:latest, llava, bakllava, llava-phi3:latest  and gemma3:4b-it-qat.
It is required to only pull reuired models within ollama container before starting the vision container.


Unavailable ollama should return:
```json
{
  "error": "Ollama processing error: Failed to connect to Ollama. Please check that Ollama is downloaded, running and accessible. https://ollama.com/download"
}
```

**Comment by lastzero** [comment]
> I have following configuration:

Thanks for sharing! This should help. 👍

Did you compare the output and performance of the available models before you decided to use `minicpm` instead of `gemma3`? 🤔 

I did not check the model list at [ollama.com/library](https://ollama.com/library) before, only the one in our `REAMDME.md` file and in the `README.md` file of the Ollama repository, so I didn't even know that `minicpm` was available...

> Maybe it would be better to parse model and version only from JSON body, because if model contains `/` it will fail on some parsing error.

For the way we use it now, definitely. When the endpoints were first specified and developed, there was no client, so the idea was to allow testing with simple, hand-written GET requests. That's why the model name was/is in the URL, not in the JSON request body (although we wanted to support that as well, but it doesn't look like the work has been completed based on my testing and understanding of the code).

> I have used minicpm-v:latest, llava, bakllava, llava-phi3:latest and gemma3:4b-it-qat. It is required to only pull reuired models within ollama container before starting the vision container.

Okay, so running `docker exec -it ollama ollama run [model]` after starting the ollama container should be all that's needed? Is there a way to automate this without a hack, like an environment variable or some cmd flag/argument? I just did a quick check and didn't see anything obvious in the docs/examples.

Feel free to change the `vision/caption` endpoint (and any other endpoints where it makes sense) to get the `model` name and optionally the `version` from the JSON request body, instead of defaulting to the `kosmos-3` model. Requests shouldn't fail if no version is provided though, and having a default model would make sense if no model name is provided either? The other (sub-)endpoints don't necessarily have to be removed, although it might make the code easier to maintain, and they probably aren't needed anymore. So it's up to you.

I'll try to finish testing and documentation as soon as possible, so we can build a working Docker image for others to test and provide feedback! 🗯

**Comment by sgflt** [comment]
I went to conclusion, that minicpm provides better description with decent performance. But it is subjective.
Users should select model that fits their needs based on their content and requirements.


ollama models automation) I will try if ollama downloads any requested model automatically.

**Comment by sgflt** [comment]
No, it will not automatically download model as admins usually do not want user to be able to spam disk space by requesting any downloadable model.

So ollama pull have to be done by admin of ollama instance in advance.

**Comment by lastzero** [comment]
@sgflt I didn't mean to download Ollama models via REST API (remotely), but right after the `ollama` service is started. 

Ideally, we can add an additional command or flag to the `compose.yaml` file so that no manual step is required to download a model by running `docker exec -it ollama ollama run [model]` in a terminal?

**Comment by lastzero** [comment]
@sgflt I'm happy to report that I finally got my `vision.yml` properly configured so that captions can be generated with `photoprism vision run -m caption [filter]` when **(a)** the `ollama` and `photoprism-vision` services are started and **(b)** the requested Ollama model has been downloaded with `docker exec -it ollama ollama run [model]`. That's awesome! 🌈 💎 

As the `nvidia-smi` command showed, inference was successfully accelerated with my NVIDIA RTX 4060 GPU, although it still doesn't feel fast enough to let PhotoPrism generate captions with it by default due to the time and resources required. ⏳ 

**One thing to consider, though, is changing the prompt (or model response handling) so that captions don't start with "Here's a detailed description of the image:...", "This image...", or "An image of..."?** 🤔

For the intended use, it seems better if the text just describes the content of the image (and is not extremely long or markdown formatted, see screenshot below). For example:

- *A cat sleeping with its head resting on the strings of an instrument.*
- *A vibrant, full poppy flower in rich shades of red and pink.*
- *A young woman, likely in her late 20s or early 30s. She is smiling broadly, appearing friendly and approachable.*

Of course, the caption doesn't have to start with "A" and can be longer than these examples.

**To illustrate, here is a screenshot of the captions generated by the `gemma3` model with the current prompt:**

![Image](https://github.com/user-attachments/assets/63786c6c-230c-472b-994f-0e4af5725e92)

**Comment by lastzero** [comment]
I've added a new issue for improving the default caption prompts and allowing a custom prompt to be submitted:

- https://github.com/photoprism/photoprism-vision/issues/11
